### PR TITLE
Bug 1465987 - Convert the BugFiler to ReactJS

### PR DESCRIPTION
**Pull request by camd**
[Bug 1465987](https://bugzilla.mozilla.org/show_bug.cgi?id=1465987)

This is a straight over conversion.  I had to re-write quite a bit of the ``BugFiler`` but tried to keep at least some of the structure of functions the same.

A big chunk of this change is the tests.  I had to modify them to from Angular to React, of course, but the content of what they test didn't change.  Still the same input and expected outputs.  The only new test is the last one.  So you can largely skip reviewing them in detail, if you like.  :)

Testing
-------
This calls to the Treeherder API which then, in turn, ``POST``s to the bug to bugzilla.  So to test, I stubbed out the back-end to just print out the payload ``JSON`` that would be sent to BZ.  Then I created a bug with the original code and then this new code and compared the payloads to ensure they were identical.

### CODE REVIEW

File: ui/job-view/details/tabs/autoclassify/ErrorLine.jsx [hunk 1]
```diff
@@ -144,7 +144,7 @@ export default class ErrorLine extends React.Component {
    */
   onOptionChange(option) {
     this.initOption(option);
-    this.setState({ selectedOption: option });
```
Comment (camd) [inline ui/job-view/details/tabs/autoclassify/ErrorLine.jsx]: This was actually an older bug I'd missed.  When setting state, you need to set it to a new ``Object`` not just the same object that had a field modified, as we did in this case.  So doing the spread remedies that.

**Comment by camd** [comment]
@sarah-clements : Since Ed's on PTO, you're the only one on board to review this.  Sorry it's another whopper, but it's fairly contained.  This functionality is only available in two places:

1. When a failed job is selected, there is a little "Bug" icon that you can click to pull up this dialog next to each failed line.
2. In the ``AutoclassifyTab`` there's a button that looks the same.  But to enable that tab, you have to add ``&autoclassify`` to the url and reload the page.  Then when you select a failure, you'll see the new tab.  

To be honest, since the ``AutoclassifyTab`` is disabled by default and no one is using it, you don't need to worry about it too much.  When we decide to re-enable it in some capacity after more of George's work (we may do away with it altogether and incorporate his work into the normal ``Failure Summary`` tab anyway...) then we'll go through the UI with more pointed testing.  But your call there.

File: ui/job-view/details/BugFiler.jsx [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+      <div>
+        <Modal isOpen={isOpen} toggle={toggle} size="lg">
+          <ModalHeader toggle={toggle}>Intermittent Bug Filer</ModalHeader>
+          <ModalBody>
+            <form id="modalForm">
+              <input
+                name="modalProductFinderSearch"
+                id="modalProductFinderSearch"
+                onKeyDown={this.productSearchEnter}
+                onChange={evt => this.setState({ productSearch: evt.target.value })}
```
Comment (sarah-clements) [inline ui/job-view/details/BugFiler.jsx]: I thought calling setState in render was problematic due to the state/render cycle? Is that not the case if its in-lined in jsx? (seems to work fine when testing)

File: ui/css/treeherder-bugfiler.css [hunk 3]
```diff
@@ -5,6 +5,23 @@
     padding: 20px;
 }
 
+#modalProductFinderButton {
+    margin: -4px 0 0 5px;
+    line-height: 16px;
+}
+
+#suggestedProducts {
```
Comment (sarah-clements) [inline ui/css/treeherder-bugfiler.css]: These margin attributes seem quite precise, but I wonder if you couldn't use some of the bootstrap utility spacing classes for some or if it'd be worth making them into general spacing classes to reuse?

Comment (camd) [reply ui/css/treeherder-bugfiler.css]: Yeah, this is a really good point.  I spent some time laying this out with just the bootstrap classes.  Very educational.  :)  But I eliminated most of the CSS file in the process.

File: ui/job-view/details/tabs/autoclassify/LineOption.jsx [hunk 4]
```diff
@@ -182,6 +173,18 @@ export default class LineOption extends React.Component {
...
           </span>))}
         </div>}
+        {isBugFilerOpen && <BugFiler
+          isOpen={isBugFilerOpen}
+          toggle={this.toggleBugFiler}
+          suggestion={errorLine.data.bug_suggestions}
+          suggestions={[errorLine.data.bug_suggestions]}
+          fullLog={logUrl}
+          parsedLog={`${location.origin}/${getLogViewerUrl(job.id, repoName)}`}
+          reftestUrl={isReftest(job) ? `${getReftestUrl(logUrl)}&only_show_unexpected=1` : ''}
```
[the remark this answers is not in the document]
Comment (camd) [reply ui/job-view/details/tabs/autoclassify/LineOption.jsx]: Heh, ends up that param was already in the helper.  So these were adding it twice.  Just had to remove them.

File: ui/helpers/url.js [hunk 5]
```diff
@@ -133,3 +133,9 @@ export const getRepoUrl = function getRepoUrl(newRepoName) {
   params.set('repo', newRepoName);
   return `${uiJobsUrlBase}?${params.toString()}`;
 };
+
+export const bzBaseUrl = 'https://bugzilla.mozilla.org/';
```
[the remark this answers is not in the document]
Comment (camd) [reply ui/helpers/url.js]: Yeah, I'm certain a good audit/cleanup will be in order on the helpers after all this.  We're getting close!  :)  I wish there was a better name than ``helpers``.  I almost prefer ``lib`` but not really sure.  Still kind of meaninglessly generic...  Names are hard..

File: ui/job-view/details/tabs/autoclassify/LineOption.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (edmorley) [reply ui/job-view/details/tabs/autoclassify/LineOption.jsx]: Maybe "utils" or else just put them in "shared"?

**Pull request merged**